File: src/lib/operations/branch.js
import {REG} from "../syntax/registers.js";
import state from "../state.js";
import {asm, asmBranch, asmNOP} from "../intsructions/asmWriter.js";
import {isSigned, u32InS16Range} from "../dataTypes/dataTypes.js";
import opsScalar from "./scalar.js";

// "beq", "bne", "bgezal", "bltzal", "bgez", "bltz"
const BRANCH_INVERT = {
  "beq": "bne",
  "bne": "beq",
  "bgezal": "bltzal",
  "bltzal": "bgezal",
  "bgez": "bltz",
  "bltz": "bgez",
};

export function invertBranchOp(op) {
  if(!BRANCH_INVERT[op])state.throwError("Cannot invert branch operation: " + op);
  return BRANCH_INVERT[op];
}

/**
 * Creates a branch instruction with a comparison against a register or immediate.
 * Immediate values are loaded into a register first if necessary.
 * This will only jump if the comparison fails, so the "true" case is expected to follow the branch.
 *
 * @param {ASTCompare} compare operation and left/right values
 * @param labelElse label to jump to if the comparison fails (aka "else")
 * @returns {ASM[]} list of ASM instructions
 */
export function opBranch(compare, labelElse, invert = false)
{
  compare = structuredClone(compare); // gets modified

  let isImmediate = compare.right.type === "num";
  let regTestRes = isImmediate ? REG.AT
    : state.getRequiredVar(compare.right.value, "compare").reg;

  // use register for zero-checks (avoids potential imm. load instructions)
  if(isImmediate && compare.right.value === 0) {
    isImmediate = false;
    regTestRes = REG.ZERO;
  }

  let {reg: regLeft, type: baseType} = state.getRequiredVar(compare.left.value, "left");

  // Easy case, just compare
  if(compare.op === "==" || compare.op === "!=")
  {
    let opBranch = compare.op === "==" ? "bne" : "beq";
    if(invert)opBranch = invertBranchOp(opBranch);

    return [
      ...(isImmediate ? opsScalar.loadImmediate(REG.AT, compare.right.value) : []),
      asmBranch(opBranch, [regLeft, regTestRes, labelElse], labelElse),
      asmNOP(),
    ];
  }

  const opsLoad = [];
  let regOrValRight = isImmediate ? compare.right.value : regTestRes;
  let isZeroCompare = !isImmediate && regTestRes === REG.ZERO;

  // Both ">" and "<=" are causing the biggest issues when inverted, so map them to the other two
  if(compare.op === ">" || compare.op === "<=") {
    if(isImmediate) {
      // add one to the immediate to add/remove the "=" part of the comparison.
      // Ignore overflows here (e.g. "x > 0xFFFFFFFF" would be stupid anyway)
      regOrValRight = (regOrValRight+1) >>> 0;
      compare.op = compare.op === ">" ? ">=" : "<";
    } else {
      compare.op = compare.op === ">" ? "<" : ">="; // invert comparison ...
      [regLeft, regOrValRight] = [regOrValRight, regLeft]; //... and swap registers
    }
  }

  // All values from now on need signedness checks, first check if it can still be an immediate
  if(isImmediate && !u32InS16Range(regOrValRight >>> 0)) {
    // ...if it doesn't we load it and switch back to a register check
    opsLoad.push(...opsScalar.loadImmediate(REG.AT, regOrValRight));
    isImmediate = false;
    regOrValRight = REG.AT;
  }

  const signed = isSigned(baseType); // Note: both the signed/unsigned 'slt' have a sign-extended immediate
  const opLessThan = "slt" + (isImmediate ? "i" : "") + (signed ? "" : "u");

  if(compare.op === "<" || compare.op === ">=")
  {
    let opBranch = compare.op === "<" ? "beq" : "bne";
    if(invert)opBranch = invertBranchOp(opBranch);

    if(isZeroCompare) {
      console.log("Zero compare", invert);
      opBranch = compare.op === "<" ? "bgez" : "bltz";
      if(invert)opBranch = invertBranchOp(opBranch);
      return [
        asmBranch(opBranch, [regLeft, labelElse], labelElse), // jump if "<" fails (aka ">=")
        asmNOP(),
      ];
    }

    return [
      ...opsLoad,
      asm(opLessThan, [REG.AT, regLeft, regOrValRight]),
      asmBranch(opBranch, [REG.AT, REG.ZERO, labelElse], labelElse), // jump if "<" fails (aka ">=")
      asmNOP(),
    ];
  }

  return state.throwError("Unknown comparison operator: " + compare.op, compare);
}

File: cpp/src/operations/branch.cpp
#include "branch.h"
#include "scalar.h"

#include "../asm.h"
#include "../registers.h"
#include "../state.h"
#include "../types.h"

#include <string>
#include <unordered_map>

namespace rspl::ops {

static const std::unordered_map<Opcode, Opcode> BRANCH_INVERT = []() {
  std::unordered_map<Opcode, Opcode> m;
  m[getOpcode("beq")] = getOpcode("bne");
  m[getOpcode("bne")] = getOpcode("beq");
  m[getOpcode("bgezal")] = getOpcode("bltzal");
  m[getOpcode("bltzal")] = getOpcode("bgezal");
  m[getOpcode("bgez")] = getOpcode("bltz");
  m[getOpcode("bltz")] = getOpcode("bgez");
  m[getOpcode("blez")] = getOpcode("bgtz");
  m[getOpcode("bgtz")] = getOpcode("blez");
  return m;
}();

static const std::unordered_map<std::string, Opcode>
    ZERO_COMB_BRANCH = []() {
      std::unordered_map<std::string, Opcode> m;
      m["<"] = getOpcode("bltz");
      m["<="] = getOpcode("blez");
      m[">"] = getOpcode("bgtz");
      m[">="] = getOpcode("bgez");
      return m;
    }();

Opcode invertBranchOp(Opcode op) {
  auto it = BRANCH_INVERT.find(op);
  if (it == BRANCH_INVERT.end()) {
    state.throwError("Cannot invert branch operation: " +
                     getOpcodeName(op));
  }
  return it->second;
}

std::vector<AsmInst> opBranch(const ast::CompareExpr &compare,
                              const std::string &labelElse,
                              bool invert) {
  bool isImmediate = (compare.right.type == ArgType::Num);
  std::string regTestRes;
  if (isImmediate) {
    regTestRes = reg::Reg::AT;
  } else {
    const VarDef *var =
        state.getRequiredVar(compare.right.value, "compare");
    regTestRes = var->reg;
  }

  // Zero-checks can use $zero directly
  if (isImmediate && compare.right.value == "0") {
    isImmediate = false;
    regTestRes = reg::Reg::ZERO;
  }

  VarDef varLeft =
      state.getRequiredVarCopy(compare.left.value, "left");
  std::string regLeft = varLeft.reg;

  // == and != are simple
  if (compare.op == "==" || compare.op == "!=") {
    Opcode opBranch = compare.op == "==" ? getOpcode("bne") : getOpcode("beq");
    if (invert) opBranch = invertBranchOp(opBranch);

    std::vector<AsmInst> res;
    if (isImmediate) {
      auto load = loadImmediate(reg::Reg::AT, compare.right.value);
      res.insert(res.end(), load.begin(), load.end());
    }
    res.push_back(
        asmBranch(getOpcodeName(opBranch), {regLeft, regTestRes, labelElse}, labelElse));
    res.push_back(asmNOP());
    return res;
  }

  // Zero-combination branch ops
  if (!isImmediate && regTestRes == reg::Reg::ZERO) {
    auto it = ZERO_COMB_BRANCH.find(compare.op);
    if (it != ZERO_COMB_BRANCH.end()) {
      Opcode op = it->second;
      if (!invert) op = invertBranchOp(op);
      return {asmBranch(getOpcodeName(op), {regLeft, labelElse}, labelElse), asmNOP()};
    }
  }

  // Transform > and <= into < and >= with swapped args
  std::string op = compare.op;
  std::string valR = isImmediate ? compare.right.value : regTestRes;

  if (op == ">" || op == "<=") {
    if (isImmediate) {
      // Increment immediate to handle the "=" part
      valR = std::to_string(std::stoll(valR) + 1);
      op = op == ">" ? ">=" : "<";
    } else {
      op = op == ">" ? "<" : ">=";
      std::swap(regLeft, regTestRes);
    }
  }

  // If immediate doesn't fit in 16-bit signed range, load it into $at first
  if (isImmediate) {
    int64_t immVal = std::stoll(valR);
    if (immVal < -32768 || immVal > 32767) {
      isImmediate = false;
      regTestRes = reg::Reg::AT;
    }
  }

  // For </>= comparisons, use slt + branch
  std::string opLessThan =
      "slt" + (isImmediate ? std::string("i") : "") +
      (isSigned(varLeft.type) ? "" : "u");

  if (op == "<" || op == ">=") {
    Opcode brOp = op == "<" ? getOpcode("beq") : getOpcode("bne");
    if (invert) brOp = invertBranchOp(brOp);

    std::vector<AsmInst> res;
    if (!isImmediate && regTestRes == reg::Reg::AT) {
      auto load = loadImmediate(reg::Reg::AT, valR);
      res.insert(res.end(), load.begin(), load.end());
    }
    res.push_back(
        asmOp(opLessThan, {reg::Reg::AT, regLeft,
                            isImmediate ? valR : regTestRes}));
    res.push_back(asmBranch(
        getOpcodeName(brOp), {reg::Reg::AT, reg::Reg::ZERO, labelElse}, labelElse));
    res.push_back(asmNOP());
    return res;
  }

  state.throwError(
      "Unknown comparison operator: " + compare.op, {});
  return {};
}

} // namespace rspl::ops
